## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 18
---------------------------------------

GOALS:

1. Practice Logistic Regression
2. Interpret Logistic Regression Results

----------------------------------------------------------


This homework has **1 Exercise** and **1 Challenge Exercise**

### Important Information

- Email: [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
- Office Hours: Duke 209 <a href="https://joannabieri.com/schedule.html"> Click Here for Joanna's Schedule</a>


### Announcements

**Come to Lab!** If you need help we are here to help!

### Day 18 Assignment - same drill.


1. Make sure **Pull** any new content from the class repo - then **Copy** it over into your working diretory.
2. Open the file Day##-HW.ipynb and start doing the problems.
    * You can do these problems as you follow along with the lecture notes and video.
3. Get as far as you can before class.
4. Submit what you have so far **Commit** and **Push** to Git.
5. Take the daily check in quiz on **Canvas**.
7. Come to class with lots of questions!


In [49]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.default = 'colab'

from itables import show

# This stops a few warning messages from showing
pd.options.mode.chained_assignment = None 
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

# Machine Learning Packages
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LinearRegression, LogisticRegression 
from sklearn import metrics

### Data: A collection of Emails

- Emails for the first three months of 2012 for an email account
- Data from 3921 emails and 21 variables on them
- Outcome: whether the email is spam or not
- Predictors: number of characters, whether the email had "Re:" in the subject, time at which email was sent, number of times the word "inherit" shows up in the email, etc.


Data Information: https://www.openintro.org/data/index.php?data=email

This lab follows the Data Science in a Box units "Unit 4 - Deck 6: Logistic regression" by Mine Çetinkaya-Rundel. It has been updated for our class and translated to Python by Joanna Bieri.

In [50]:
file_name = 'data/email.csv'
DF = pd.read_csv(file_name)

In [51]:
DF

,spam,to_multiple,from,cc,sent_email,time,image,attach,dollar,winner,...,viagra,password,num_char,line_breaks,format,re_subj,exclaim_subj,urgent_subj,exclaim_mess,number
0,0,0,1,0,0,2012-01-01T06:16:41Z,0,0,0,no,...,0,0,11.370,202,1,0,0,0,0,big
1,0,0,1,0,0,2012-01-01T07:03:59Z,0,0,0,no,...,0,0,10.504,202,1,0,0,0,1,small
2,0,0,1,0,0,2012-01-01T16:00:32Z,0,0,4,no,...,0,0,7.773,192,1,0,0,0,6,small
3,0,0,1,0,0,2012-01-01T09:09:49Z,0,0,0,no,...,0,0,13.256,255,1,0,0,0,48,small
4,0,0,1,0,0,2012-01-01T10:00:01Z,0,0,0,no,...,0,2,1.231,29,0,0,0,0,1,none
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3916,1,0,1,0,0,2012-03-31T00:03:45Z,0,0,0,no,...,0,0,0.332,12,0,0,0,0,0,small
3917,1,0,1,0,0,2012-03-31T14:13:19Z,0,0,1,no,...,0,0,0.323,15,0,0,0,0,0,small
3918,0,1,1,0,0,2012-03-30T16:20:33Z,0,0,0,no,...,0,0,8.656,208,1,0,0,0,5,small
3919,0,1,1,0,0,2012-03-28T16:00:49Z,0,0,0,no,...,0,0,10.185,132,0,0,0,0,0,small


In [52]:
DF.shape

(3921, 21)

In [53]:
DF.columns

Index(['spam', 'to_multiple', 'from', 'cc', 'sent_email', 'time', 'image',
       'attach', 'dollar', 'winner', 'inherit', 'viagra', 'password',
       'num_char', 'line_breaks', 'format', 're_subj', 'exclaim_subj',
       'urgent_subj', 'exclaim_mess', 'number'],
      dtype='object')

In [54]:
DF.describe()

,spam,to_multiple,from,cc,sent_email,image,attach,dollar,inherit,viagra,password,num_char,line_breaks,format,re_subj,exclaim_subj,urgent_subj,exclaim_mess
count,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000,3921.000000
mean,0.093599,0.158123,0.999235,0.404489,0.277990,0.048457,0.132874,1.467228,0.038001,0.002040,0.108136,10.706586,230.658505,0.695231,0.261413,0.080337,0.001785,6.584290
std,0.291307,0.364903,0.027654,2.666424,0.448066,0.450848,0.718518,5.022298,0.267899,0.127759,0.959931,14.645786,319.304959,0.460368,0.439460,0.271848,0.042220,51.479871
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.001000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.459000,34.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,5.856000,119.000000,1.000000,0.000000,0.000000,0.000000,1.000000
75%,0.000000,0.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.084000,298.000000,1.000000,1.000000,0.000000,0.000000,4.000000
max,1.000000,1.000000,1.000000,68.000000,1.000000,20.000000,21.000000,64.000000,9.000000,8.000000,28.000000,190.087000,4022.000000,1.000000,1.000000,1.000000,1.000000,1236.000000


In [55]:
DF.dtypes

spam              int64
to_multiple       int64
from              int64
cc                int64
sent_email        int64
time             object
image             int64
attach            int64
dollar            int64
winner           object
inherit           int64
viagra            int64
password          int64
num_char        float64
line_breaks       int64
format            int64
re_subj           int64
exclaim_subj      int64
urgent_subj       int64
exclaim_mess      int64
number           object
dtype: object

In [56]:
DF['spam'].value_counts()

spam
0    3554
1     367
Name: count, dtype: int64

i would think that the emails containing winner, viagra, dollar, urgent_subj would be more likely to be spam

i would think spam emails would be shorter and to the point so people can quickly read what they are trying to get across

the emails i have gotten that are legit say RE so maybe the other ones are not reliable and more likely spam

**Exercise 1** Logistic Regression with ONE explanatory variable.

Choose another variable from the data set to use as your explanatory variable and create a Logistic Regression model to predict if an email is spam or not. You should do all of the following:

1. Say what variable you are using to predict spam messages (do some analysis, at minimum a value_counts()). Why do you think this is a good variable to use in predicting if an email is spam.
2. Create and fit a Logistic Regression model.
3. Show the results: intercept, coefficient, basic confusion matrix prediction.
4. What do you think the decision cutoff should be? Update the cutoff and redo the confusion matrix.
5. Explain your results in words. You should talk about False Negative and False positive rates and what they mean in terms of the variables you chose.


**Exercise 2 - challenge** Logistic Regression with MORE THAN ONE explanatory variable.

Try redoing the analysis, but this time add a few more explanatory variables. Again do some analysis of the variables you are chosing and state why they are a good choice. Then answer again questions 1-5.

In [106]:
DF[['line_breaks','spam']].groupby('spam').describe()

line_breaks                                                         
           count        mean         std  min   25%    50%    75%     max
spam                                                                     
0         3554.0  243.653067  318.257105  2.0  42.0  137.0  315.0  4022.0
1          367.0  104.820163  301.995685  1.0  14.0   22.0   65.5  3729.0

In [71]:
# 1. Why do you think this is a good variable to use in predicting if an
# email is spam

this is a good variable perhaps because the max in the spam has quite a spike and definetly a different spread than the non spam emails so there is definetly some variability there. I feel that emails with a suspicious amount of extra spacing, or possibly too little, are sketchy because spam email senders tend to mess with the line formatting to make their emails seem more legit.

In [59]:
# 2. Create and fit a Logistic Regression model.

In [97]:
# Get a subset of the rows
DF_model = DF[['line_breaks','spam']]

# Get the variables
X = DF_model['line_breaks'].values.reshape(-1,1)
y = DF_model['spam']

# Do the regression
LM = LogisticRegression()
LM.fit(X,y)

print('Classes:')
print(LM.classes_)
print('Coefficients:')
print(LM.coef_)
print('Intercept:')
print(LM.intercept_)

Classes:
[0 1]
Coefficients:
[[-0.00345463]]
Intercept:
[-1.73910342]


In [98]:
num = 2
intercept = -1.74
slope = -0.0035

eta = intercept + slope*num

P = np.exp(eta)/(1+np.exp(eta))
print(P)

0.1484259853689531


In [99]:
num = 40
intercept = -1.74
slope = -0.0035

eta = intercept + slope*num

P = np.exp(eta)/(1+np.exp(eta))
print(P)

0.1323888735420654


In [100]:
intercept = -1.74
slope = -0.0035

P = []

num_characters = list(np.arange(0,200,10))

for num in num_characters:
    eta = intercept + slope*num
    P.append(np.exp(eta)/(1+np.exp(eta)))


# Plot the results
fig = px.scatter(DF_model,x='line_breaks',y='spam',opacity=.5)

fig.add_trace(
    px.line(x=num_characters,y=P,color_discrete_sequence=['black']).data[0]
)

fig.show()

print(LM.intercept_)
print(LM.coef_)

[-1.73910342]
[[-0.00345463]]


In [101]:
# 3. Show the results: intercept, coefficient, basic confusion matrix
# prediction.

# import the metrics class
from sklearn import metrics

DF_model['prediction'] = LM.predict(X)

y_pred = DF_model['prediction'].values

cnf_matrix = metrics.confusion_matrix(y, y_pred)
cnf_matrix

array([[3554,    0],
       [ 367,    0]])

3554 true positives, 367 false negatives, 0 false positives, 0 true negatives

In [102]:
# 4. What do you think the decision cutoff should be? Update the cutoff
# and redo the confusion matrix.

# Create new columns in your data frame
DF_model[['prob not spam','prob spam']] = LM.predict_proba(X)
# Look at the model DF
DF_model

,line_breaks,spam,prediction,prob not spam,prob spam
0,202,0,0,0.919601,0.080399
1,202,0,0,0.919601,0.080399
2,192,0,0,0.917010,0.082990
3,255,0,0,0.932140,0.067860
4,29,0,0,0.862864,0.137136
...,...,...,...,...,...
3916,12,1,0,0.855766,0.144234
3917,15,1,0,0.857040,0.142960
3918,208,0,0,0.921120,0.078880
3919,132,0,0,0.899810,0.100190


In [103]:
cutoff = .1 #if prob spam is 5% or higher it is labeled spam

# Use a lambda on the new columns to create your new prediction
DF_model['new_prediction'] = DF_model['prob spam'].apply(lambda x: 1 if x>cutoff else 0)

# Look at the model DF
DF_model

,line_breaks,spam,prediction,prob not spam,prob spam,new_prediction
0,202,0,0,0.919601,0.080399,0
1,202,0,0,0.919601,0.080399,0
2,192,0,0,0.917010,0.082990,0
3,255,0,0,0.932140,0.067860,0
4,29,0,0,0.862864,0.137136,1
...,...,...,...,...,...,...
3916,12,1,0,0.855766,0.144234,1
3917,15,1,0,0.857040,0.142960,1
3918,208,0,0,0.921120,0.078880,0
3919,132,0,0,0.899810,0.100190,1


In [104]:
# Redo the confusion matrix

y_pred = DF_model['new_prediction'].values

cnf_matrix = metrics.confusion_matrix(y, y_pred)
cnf_matrix

array([[1820, 1734],
       [  65,  302]])

In [105]:
TP = cnf_matrix[0,0]
FP = cnf_matrix[0,1]
FN = cnf_matrix[1,0]
TN = cnf_matrix[1,1]

print('False Negative Rate:')
print(FN/ (TP+FN))
print('------')
print('False Positive Rate:')
print(FP/ (FP+TN))
print('------')
print('Sensitivity:')
print(1 - (FN/ (TP+FN)))
print('------')
print('Specificity:')
print(1- (FP/ (FP+TN)))

False Negative Rate:
0.034482758620689655
------
False Positive Rate:
0.8516699410609038
------
Sensitivity:
0.9655172413793104
------
Specificity:
0.14833005893909623


In [69]:
# 5. Explain your results in words. You should talk about False Negative
# and False positive rates and what they mean in terms of the variables
# you chose.

False negative means an email not spam and it was labeled spam. This happened in 3.4% of cases where the email was not spam.

False positive means that am email was spam and was labeled not spam. This happened in 85.2% of the cases where the email was spam.

If the email was not spam it has a 96.5% probability of being labeled not spam.

If the email was spam it has a 14.8% probability of being labeled spam.

These results are not much different from the number of characters results.